clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java

```diff
@@ -2999,14 +3004,32 @@ private int processKafkaDataMessage(
         keyLen = keyBytes.length;
         // update checksum for this PUT message if needed.
         partitionConsumptionState.maybeUpdateExpectedChecksum(keyBytes, put);
-        prependHeaderAndWriteToStorageEngine(
-            // Leaders might consume from a RT topic and immediately write into StorageEngine,
-            // so we need to re-calculate partition.
-            // Followers are not affected since they are always consuming from VTs.
-            producedPartition,
-            keyBytes,
-            put,
-            currentTimeMs);
+
+        // Do transorfmation recompute key, value and partition
+        if (recordTransformer != null) {
+          Lazy<byte[]> lazyKey = Lazy.of(() -> keyBytes);
+          Lazy<ByteBuffer> lazyValue = Lazy.of(() -> put.putValue);
+          GenericRecord transformedRecord = recordTransformer.put(lazyKey, lazyValue, producedPartition);
+          ByteBuffer keyByteBuffer = (ByteBuffer) transformedRecord.get("key");
+          byte[] newKeyBytes = new byte[keyByteBuffer.remaining()];
+          keyByteBuffer.get(newKeyBytes);
+          ByteBuffer valueByteBuffer = (ByteBuffer) transformedRecord.get("value");
+          byte[] valueBytes = new byte[valueByteBuffer.remaining()];
+          valueByteBuffer.get(valueBytes);
+          int newProducedPartition = Objects.hash(newKeyBytes, put.putValue);
```

Comment:
This isn't quite right.  Use VenicePartitioner.getPartitionId(keybytes, numPartitions).  Theres already a VenicePartitioner member of StoreIngestionTask, so should be an easy tweak.